// @nest/core/exceptions/exceptions-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const invalid_exception_filter_exception_1 = require("../errors/exceptions/invalid-exception-filter.exception");
const http_exception_1 = require("../exceptions/http.exception");
const logger_service_1 = require("../services/logger.service");
const shared_utils_1 = require("../utils/shared.utils");
const http_exception_2 = require("./http-exception");
class ExceptionsHandler {
    constructor() {
        this.filters = [];
    }
    next(exception, response) {
        if (this.invokeCustomFilters(exception, response))
            return;
        if (!(exception instanceof http_exception_1.HttpException ||
            exception instanceof http_exception_2.HttpException)) {
            response.status(500).json({
                statusCode: 500,
                message: constants_1.messages.UNKNOWN_EXCEPTION_MESSAGE,
            });
            if (shared_utils_1.isObject(exception) && exception.message) {
                return ExceptionsHandler.logger.error(exception.message, exception.stack);
            }
            return ExceptionsHandler.logger.error(exception);
        }
        const res = exception.getResponse();
        const message = shared_utils_1.isObject(res)
            ? res
            : {
                statusCode: exception.getStatus(),
                message: res,
            };
        response.status(exception.getStatus()).json(message);
    }
    setCustomFilters(filters) {
        if (!Array.isArray(filters)) {
            throw new invalid_exception_filter_exception_1.InvalidExceptionFilterException();
        }
        this.filters = filters;
    }
    invokeCustomFilters(exception, response) {
        if (shared_utils_1.isEmpty(this.filters))
            return false;
        const filter = this.filters.find(({ exceptionMetatypes, func }) => {
            const hasMetatype = !exceptionMetatypes.length ||
                !!exceptionMetatypes.find(ExceptionMetatype => exception instanceof ExceptionMetatype);
            return hasMetatype;
        });
        filter && filter.func(exception, response);
        return !!filter;
    }
}
ExceptionsHandler.logger = new logger_service_1.Logger(ExceptionsHandler.name);
exports.ExceptionsHandler = ExceptionsHandler;
